jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java

```diff
@@ -596,7 +598,40 @@ default Enumeration<String> getFieldNames()
      */
     default Set<String> getFieldNamesCollection()
     {
-        return stream().map(HttpField::getName).collect(Collectors.toSet());
+        Set<HttpHeader> seenByHeader = EnumSet.noneOf(HttpHeader.class);
+        Set<String> seenByName = null;
+        List<String> list = new ArrayList<>(size());
+
+        for (HttpField f : this)
+        {
+            HttpHeader header = f.getHeader();
+            if (header == null)
+            {
+                seenByName = Objects.requireNonNullElse(seenByName, new TreeSet<>(String::compareToIgnoreCase));
```

Comment:
Looks like this creates a new `TreeSet` every time and only uses it if `seenByName == null`?

Maybe something like this is better?
```java
if (seenByName == null)
    seenByName = new TreeSet<>(String::compareToIgnoreCase);
```

Reply:
ooops